src/host/acme_challenge_server.h

```diff
@@ -0,0 +1,283 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ccf/ds/hex.h"
+#include "ccf/ds/logger.h"
+#include "ccf/http_status.h"
+#include "ccf/json_handler.h"
+#include "ds/cli_helper.h"
+#include "ds/ring_buffer_types.h"
+#include "enclave/interface.h"
+#include "host/socket.h"
+#include "http/http_builder.h"
+#include "http/http_endpoint.h"
+#include "http/http_parser.h"
+#include "tcp.h"
+
+#include <cstddef>
+#include <fmt/format.h>
+#include <memory>
+#include <stdexcept>
+#include <uv.h>
+
+class ACMEConnectionTracker
+{
+public:
+  virtual void add(asynchost::TCP& peer) = 0;
+};
+
+class ACMEServerBehaviour : public asynchost::SocketBehaviour<asynchost::TCP>
+{
+protected:
+  class ClientBehaviour : public asynchost::SocketBehaviour<asynchost::TCP>
+  {
+  public:
+    ClientBehaviour(
+      asynchost::TCP socket,
+      std::mutex& lock,
+      const std::map<std::string, std::string>& prepared_responses,
+      ringbuffer::WriterPtr to_enclave) :
+      asynchost::SocketBehaviour<asynchost::TCP>("", ""),
+      parser(sp),
+      socket(socket),
+      lock(lock),
+      prepared_responses(prepared_responses),
+      to_enclave(to_enclave)
+    {}
+
+    virtual ~ClientBehaviour() {}
+
+    void reply(http::Response& r, const std::string& body)
+    {
+      if (!socket.is_null())
+      {
+        std::vector<uint8_t> vody(body.begin(), body.end());
+        r.set_body(vody.data(), vody.size());
+        auto bytes = r.build_response();
+        socket->write(bytes.size(), bytes.data());
+      }
+    }
+
+    virtual void on_read(size_t len, uint8_t*& incoming, sockaddr sa) override
+    {
+      std::string body((char*)incoming, len);
```

Comment:
I may be missing something here, but it looks like we don't mutate body, so do we really need the copy?

Reply:
I just need that string when printing an error message, so I moved it further down. The reason I wanted a string is that I can't rely on `incoming` being zero-terminated.